### PR TITLE
i.emissivity: added test file for i.emissivity module

### PR DESCRIPTION
**Pull request by jayneel-shah18**
This PR introduces a comprehensive test suite for the `i.emissivity` GRASS GIS module. The test suite is designed to validate core functionality, assess statistical correctness, and handle edge cases to ensure robust module behavior.

### Key Updates in this PR:
---
- Validate raster output generation and behavior under different NDVI conditions.
- Ensure correct handling of masked regions.
- Test behavior with different spatial resolutions.
- Verify response to extreme NDVI values and partial null inputs.

### **Test Case Additions**
---
#### **Basic Functionality**
- **NDVI Range Validity:** Ensure correct emissivity values are generated for valid NDVI inputs (`test_emissivity_ndvi_range`).
- **Masked Regions:** Validate handling of irregularly shaped masks (`test_complex_mask`).

#### **Advanced Properties**
- **Spatial Resolution Handling:** Check that the module produces consistent results across different region resolutions (`test_region_resolution`).
- **Null Value Handling:** Assess behavior when NDVI contains null values (`test_partial_null_values`).
- **Extreme NDVI Values:** Test module response when NDVI is outside the valid range (`test_extreme_ndvi_values`).

#### **Performance**
- The test suite execution times:
  - **Windows:** 2.315s
  - **macOS:** 1.353s
  - **Ubuntu:** 1.868s

### Future Enhancements
---
- **Performance benchmarking** for large-scale regions.
- **Stress testing** with extreme parameter variations.

This test suite enhances the reliability of the `i.emissivity` module by covering a broad spectrum of cases. Looking forward to feedback and potential refinements!

### CODE REVIEW

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        # For col 1 (NDVI = 0.16): emissivity = 0.97
+        # For col 2 (NDVI = 0.74): emissivity = 0.99
+        # For col 3+ (NDVI = 0.45): emissivity = 0.97 + 0.003 * 0.45 = 0.97135
+        reference_stats = {
+            "mean": 0.97712,  # (0.97 + 0.99 + 0.97135) / 3
+            "min": 0.92,
+            "max": 0.99,
+        }
+        self.assertRasterFitsUnivar(
+            raster=self.output_raster, reference=reference_stats, precision=1e-2
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:73]: Why precision 1e-2?

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        self.runModule(
+            "r.mapcalc",
+            expression=f"diff_raster = abs({self.output_raster} - masked_reference)",
+            overwrite=True,
+        )
+        self.temp_rasters.append("diff_raster")
+
+        self.assertRasterMinMax("diff_raster", 0, 0.05)
+        self.runModule("r.mask", flags="r")
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:107]: You need to remove the mask before you test the outputs.

It should be enough to check with assertRasterFitsUnivar, e.g., the mean, min, max and non-null cells.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                )
+                self.assertModule(
+                    "i.emissivity",
+                    input=self.input_raster,
+                    output=self.output_raster,
+                    overwrite=True,
+                )
+                results[res] = self.assertRasterMinMax(self.output_raster, 0, 1)
+
+        self.assertAlmostEqual(results[1], results[0.1], delta=0.01)
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:126]: I don't see much value in this test for i.emissivity specifically

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            overwrite=True,
+        )
+        self.assertModule(
+            "i.emissivity",
+            input=self.input_raster,
+            output=self.output_raster,
+            overwrite=True,
+        )
+        self.assertRasterExists(self.output_raster)
+        self.assertRasterMinMax(self.output_raster, 0, 1)
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:142]: I would test with assertRasterFitsUnivar

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        cls.reference_raster = "emissivity_reference_map"
+        cls.temp_rasters = []
+
+        cls.runModule("g.region", n=10, s=0, e=10, w=0, rows=10, cols=10)
+
+        # Predefined NDVI test cases
+        cls.ndvi_cases = {
+            "default": "if(row() <= 5, 0.2, 0.7)",
+            "valid_range": "if(col() == 1, 0.16, if(col() == 2, 0.74, 0.45))",
+            "extreme_values": "if(col() == 1, -1, if(col() == 2, 1.2, 0.5))",
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:24]: Isn't maximum ndvi 1, not 1.2?

File: imagery/i.emissivity/testsuite/test_i_emissivity.py (hunk 1 above)
Comment (jayneel-shah18) [reply imagery/i.emissivity/testsuite/test_i_emissivity.py:73]: I chose 1e-2 because the reference stats are theoretical, not derived from the raster. Small floating-point differences are expected due to raster computations which give Assertion Error with a stricter precision. A stricter precision (e.g., 1e-6) would require using reference stats from the actual raster output.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py (hunk 5 above)
Comment (jayneel-shah18) [reply imagery/i.emissivity/testsuite/test_i_emissivity.py:24]: Yes you are right, I had mistyped the value. Thank you for catching that.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py (hunk 2 above)
Comment (jayneel-shah18) [reply imagery/i.emissivity/testsuite/test_i_emissivity.py:99]: Thanks for the feedback! I have removed the mask before testing the outputs and updated the test to use `assertRasterFitsUnivar` to check the mean, min and max as suggested.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py (hunk 3 above)
Comment (jayneel-shah18) [reply imagery/i.emissivity/testsuite/test_i_emissivity.py]: I have replaced the previous test with the new one focused on default emissivity calculation using critical NDVI values. This ensures that if the module is changed in the future, the test will catch it by failing.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py (hunk 4 above)
Comment (jayneel-shah18) [reply imagery/i.emissivity/testsuite/test_i_emissivity.py]: I have updated the test to use `assertRasterFitsUnivar` for validation.

File: imagery/i.emissivity/testsuite/test_i_emissivity.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            (0.10, 0.96),  # Below NDVI_soil - should use e_soil
+            (0.15, 0.96),  # Exactly NDVI_soil
+            (0.70, 0.985),  # Exactly NDVI_veg
+            (0.80, 0.985),  # Above NDVI_veg - should clamp to e_veg
+            (-0.5, 0.96),  # Invalid NDVI (negative) treated as soil
+            (1.50, 0.985),  # Invalid NDVI (>1) treated as vegetation
+        ]
+
+        for ndvi, expected in test_values:
+            with self.subTest(ndvi=ndvi, expected=expected):
```
Comment (petrasovaa) [inline imagery/i.emissivity/testsuite/test_i_emissivity.py:120]: I think the problem in the CI is related to the usage of subtest, we don't use it anywhere else. So try getting rid of it.

**Comment by echoix** [comment]
You don't need to update the branch as much, if there's not a particular reason for it. The last merge hadn't even completed yet

**Review by petrasovaa — APPROVED** [review]
